File: Arrays/fourNumberSum.js
const fourNumberSum = (array, targetSum) => {
    const allPairs = {}
    const quadruplets = []

    for (let i = 1; i < array.length - 1; i++) {
        for (let j = i + 1; j < array.length; j++) {
            const currentSum = array[i] + array[j]
            const difference = targetSum - currentSum
            if (difference in allPairs) {
                for (const pair in allPairs[difference]) {
                    quadruplets.push(pair.concat([array[i], array[j]]))
                }
            }
        }
        for (let k = 0; k < i; k++) {
            const currentSum = array[i] + array[k]
            if (!currentSum in allPairs) {
                allPairs[currentSum] = [array[i], array[k]]
            } else {
                allPairs[currentSum].push([array[i], array[k]])
            }
        }
    }
    return quadruplets
}

File: Arrays/four_number_sum.py
def four_number_sum(array, targetSum):
    allPairs = {}
    quadruplets = []
    for i in range(1, len(array) - 1):
        for j in range(i + 1, len(array)):
            currentSum = array[i] + array[j]
            difference = targetSum - currentSum
            if difference in allPairs:
                for pair in allPairs[difference]:
                    quadruplets.append(pair + [array[i], array[j]])
        for k in range(0, i):
            currentSum = array[i] + array[k]
            if currentSum not in allPairs:
                allPairs[currentSum] = [[array[i], array[k]]]
            else:
                allPairs[currentSum].append([array[i], array[k]])
    return quadruplets